File: src/oauth.ts
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { decryptSecret, encryptSecret } from "./crypto.js";
import type { MailAccount } from "./types.js";

export type OAuthMailProvider = "google" | "microsoft";

type OAuthCredential = {
  version: 1;
  provider: OAuthMailProvider;
  email: string;
  displayName: string;
  clientId: string;
  accessToken: string;
  refreshToken: string;
  expiresAt: number;
  scope: string;
};

type OAuthFlow = {
  id: string;
  state: string;
  nonce: string;
  verifier: string;
  provider: OAuthMailProvider;
  redirectUri: string;
  createdAt: number;
  status: "pending" | "authorized" | "success" | "error";
  error: string;
  accountId: number | null;
};

type TokenResponse = {
  access_token?: string;
  refresh_token?: string;
  expires_in?: number;
  scope?: string;
  id_token?: string;
  error?: string;
  error_description?: string;
};

type OAuthManagerOptions = {
  encryptionKey: Buffer;
  databasePath: string;
  port: number;
  googleClientId: string;
  microsoftClientId: string;
  redirectBaseUrl?: string;
};

const flowLifetimeMs = 10 * 60_000;
const refreshSkewMs = 2 * 60_000;
const googleScopes = ["openid", "email", "profile", "https://mail.google.com/"];
const microsoftScopes = [
  "openid",
  "profile",
  "email",
  "offline_access",
  "https://outlook.office.com/IMAP.AccessAsUser.All",
  "https://outlook.office.com/SMTP.Send"
];

function base64Url(input: Buffer): string {
  return input.toString("base64url");
}

function oauthFailure(message: string, status = 401): Error {
  const error = new Error(message) as Error & { status: number };
  error.status = status;
  return error;
}

function decodeJwtPayload(token: string): Record<string, unknown> {
  const part = token.split(".")[1];
  if (!part) throw oauthFailure("OAuth 身份令牌格式不正确");
  try {
    return JSON.parse(Buffer.from(part, "base64url").toString("utf8")) as Record<string, unknown>;
  } catch {
    throw oauthFailure("OAuth 身份令牌无法解析");
  }
}

function stringClaim(payload: Record<string, unknown>, ...names: string[]): string {
  for (const name of names) {
    const value = payload[name];
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return "";
}

function audienceMatches(payload: Record<string, unknown>, clientId: string): boolean {
  const audience = payload.aud;
  if (typeof audience === "string") return audience === clientId;
  return Array.isArray(audience) && audience.some((value) => value === clientId);
}

class OAuthCredentialStore {
  constructor(private readonly filePath: string, private readonly encryptionKey: Buffer) {}

  get(syncId: string): OAuthCredential | null {
    const encrypted = this.read()[syncId];
    if (!encrypted) return null;
    try {
      const parsed = JSON.parse(decryptSecret(encrypted, this.encryptionKey)) as OAuthCredential;
      if (parsed.version !== 1 || !parsed.provider || !parsed.refreshToken) throw new Error("invalid credential");
      return parsed;
    } catch {
      throw oauthFailure("OAuth 凭据存储损坏，请重新授权邮箱");
    }
  }

  set(syncId: string, credential: OAuthCredential): void {
    const records = this.read();
    records[syncId] = encryptSecret(JSON.stringify(credential), this.encryptionKey);
    this.write(records);
  }

  delete(syncId: string): void {
    const records = this.read();
    if (!(syncId in records)) return;
    delete records[syncId];
    this.write(records);
  }

  private read(): Record<string, string> {
    if (!fs.existsSync(this.filePath)) return {};
    try {
      const parsed = JSON.parse(fs.readFileSync(this.filePath, "utf8")) as { version?: number; records?: Record<string, string> };
      if (parsed.version !== 1 || !parsed.records || typeof parsed.records !== "object") throw new Error("invalid store");
      return parsed.records;
    } catch {
      throw oauthFailure("OAuth 凭据文件无法读取，请检查本地数据目录");
    }
  }

  private write(records: Record<string, string>): void {
    fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
    fs.writeFileSync(this.filePath, `${JSON.stringify({ version: 1, records }, null, 2)}\n`, { encoding: "utf8", mode: 0o600 });
  }
}

export class OAuthManager {
  private readonly flows = new Map<string, OAuthFlow>();
  private readonly store: OAuthCredentialStore;

  constructor(private readonly options: OAuthManagerOptions) {
    this.store = new OAuthCredentialStore(`${options.databasePath}.oauth-secrets.json`, options.encryptionKey);
  }

  available(provider: OAuthMailProvider): boolean {
    return Boolean(this.clientId(provider));
  }

  marker(provider: OAuthMailProvider): string {
    return encryptSecret(`oauth-v1:${provider}`, this.options.encryptionKey);
  }

  providerForAccount(account: MailAccount): OAuthMailProvider | null {
    try {
      const value = decryptSecret(account.encryptedPassword, this.options.encryptionKey);
      if (value === "oauth-v1:google") return "google";
      if (value === "oauth-v1:microsoft") return "microsoft";
      return null;
    } catch {
      return null;
    }
  }

  saveCredential(syncId: string, credential: OAuthCredential): void {
    this.store.set(syncId, credential);
  }

  deleteCredential(syncId: string): void {
    this.store.delete(syncId);
  }

  start(provider: OAuthMailProvider): { flowId: string; authorizationUrl: string } {
    this.cleanupFlows();
    const clientId = this.clientId(provider);
    if (!clientId) {
      throw oauthFailure(provider === "google" ? "尚未配置 Google OAuth Client ID" : "尚未配置 Microsoft OAuth Client ID", 503);
    }
    const verifier = base64Url(crypto.randomBytes(48));
    const challenge = base64Url(crypto.createHash("sha256").update(verifier).digest());
    const flow: OAuthFlow = {
      id: crypto.randomUUID(),
      state: base64Url(crypto.randomBytes(32)),
      nonce: base64Url(crypto.randomBytes(24)),
      verifier,
      provider,
      redirectUri: this.redirectUri(provider),
      createdAt: Date.now(),
      status: "pending",
      error: "",
      accountId: null
    };
    this.flows.set(flow.id, flow);

    const params = new URLSearchParams({
      client_id: clientId,
      response_type: "code",
      redirect_uri: flow.redirectUri,
      response_mode: "query",
      scope: this.scopes(provider).join(" "),
      state: flow.state,
      nonce: flow.nonce,
      code_challenge: challenge,
      code_challenge_method: "S256"
    });
    if (provider === "google") {
      params.set("access_type", "offline");
      params.set("prompt", "consent select_account");
    } else {
      params.set("prompt", "select_account");
    }
    const endpoint = provider === "google"
      ? "https://accounts.google.com/o/oauth2/v2/auth"
      : "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
    return { flowId: flow.id, authorizationUrl: `${endpoint}?${params}` };
  }

  async completeCallback(state: string, code: string, providerError = "", providerErrorDescription = ""): Promise<{ flowId: string; credential: OAuthCredential }> {
    this.cleanupFlows();
    const flow = [...this.flows.values()].find((item) => item.state === state);
    if (!flow || flow.status !== "pending") throw oauthFailure("OAuth 授权请求不存在或已过期", 400);
    if (providerError) {
      const message = providerErrorDescription || providerError;
      this.markFlowError(flow.id, `授权未完成：${message}`);
      throw oauthFailure(`授权未完成：${message}`, 400);
    }
    if (!code) {
      this.markFlowError(flow.id, "授权服务器没有返回授权码");
      throw oauthFailure("授权服务器没有返回授权码", 400);
    }

    try {
      const credential = await this.exchangeCode(flow, code);
      flow.status = "authorized";
      return { flowId: flow.id, credential };
    } catch (error) {
      const message = error instanceof Error ? error.message : "OAuth 授权失败";
      this.markFlowError(flow.id, message);
      throw error;
    }
  }

  flowStatus(flowId: string): { status: OAuthFlow["status"]; error: string; accountId: number | null } {
    this.cleanupFlows();
    const flow = this.flows.get(flowId);
    if (!flow) return { status: "error", error: "授权请求已过期，请重新开始", accountId: null };
    return { status: flow.status, error: flow.error, accountId: flow.accountId };
  }

  markFlowSuccess(flowId: string, accountId: number): void {
    const flow = this.flows.get(flowId);
    if (!flow) return;
    flow.status = "success";
    flow.accountId = accountId;
    flow.error = "";
  }

  markFlowError(flowId: string, message: string): void {
    const flow = this.flows.get(flowId);
    if (!flow) return;
    flow.status = "error";
    flow.error = message.slice(0, 1000);
  }

  async accessToken(account: MailAccount): Promise<string> {
    const provider = this.providerForAccount(account);
    if (!provider) throw oauthFailure("该邮箱不是 OAuth 账户", 400);
    const credential = this.store.get(account.syncId);
    if (!credential || credential.provider !== provider) throw oauthFailure("OAuth 授权信息不存在，请重新授权邮箱");
    if (credential.accessToken && credential.expiresAt > Date.now() + refreshSkewMs) return credential.accessToken;
    const refreshed = await this.refresh(credential);
    this.store.set(account.syncId, refreshed);
    return refreshed.accessToken;
  }

  private clientId(provider: OAuthMailProvider): string {
    return (provider === "google" ? this.options.googleClientId : this.options.microsoftClientId).trim();
  }

  private scopes(provider: OAuthMailProvider): string[] {
    return provider === "google" ? googleScopes : microsoftScopes;
  }

  private redirectUri(provider: OAuthMailProvider): string {
    const configured = this.options.redirectBaseUrl?.trim().replace(/\/+$/, "");
    if (configured) return configured;
    return provider === "google"
      ? `http://127.0.0.1:${this.options.port}`
      : `http://localhost:${this.options.port}`;
  }

  private async exchangeCode(flow: OAuthFlow, code: string): Promise<OAuthCredential> {
    const clientId = this.clientId(flow.provider);
    const tokenEndpoint = flow.provider === "google"
      ? "https://oauth2.googleapis.com/token"
      : "https://login.microsoftonline.com/common/oauth2/v2.0/token";
    const body = new URLSearchParams({
      client_id: clientId,
      code,
      code_verifier: flow.verifier,
      redirect_uri: flow.redirectUri,
      grant_type: "authorization_code"
    });
    if (flow.provider === "microsoft") body.set("scope", this.scopes(flow.provider).join(" "));
    const token = await this.tokenRequest(tokenEndpoint, body);
    if (!token.access_token || !token.refresh_token || !token.id_token) {
      throw oauthFailure("授权服务器没有返回完整的 OAuth 凭据，请撤销旧授权后重试");
    }
    const claims = decodeJwtPayload(token.id_token);
    if (!audienceMatches(claims, clientId)) throw oauthFailure("OAuth 身份令牌的客户端不匹配");
    if (stringClaim(claims, "nonce") !== flow.nonce) throw oauthFailure("OAuth 身份令牌 nonce 校验失败");
    const issuer = stringClaim(claims, "iss");
    if (flow.provider === "google" && !["https://accounts.google.com", "accounts.google.com"].includes(issuer)) {
      throw oauthFailure("Google OAuth 身份令牌签发方不正确");
    }
    if (flow.provider === "microsoft" && !issuer.startsWith("https://login.microsoftonline.com/")) {
      throw oauthFailure("Microsoft OAuth 身份令牌签发方不正确");
    }
    const email = stringClaim(claims, "email", "preferred_username", "upn");
    if (!email.includes("@")) throw oauthFailure("无法从 OAuth 授权结果识别邮箱地址");
    if (flow.provider === "google" && claims.email_verified === false) throw oauthFailure("Google 账户邮箱地址尚未验证");
    return {
      version: 1,
      provider: flow.provider,
      email,
      displayName: stringClaim(claims, "name") || email.split("@")[0] || email,
      clientId,
      accessToken: token.access_token,
      refreshToken: token.refresh_token,
      expiresAt: Date.now() + Math.max(60, Number(token.expires_in ?? 3600)) * 1000,
      scope: token.scope || this.scopes(flow.provider).join(" ")
    };
  }

  private async refresh(credential: OAuthCredential): Promise<OAuthCredential> {
    const clientId = this.clientId(credential.provider);
    if (!clientId || clientId !== credential.clientId) throw oauthFailure("OAuth Client ID 已变化，请重新授权邮箱");
    const tokenEndpoint = credential.provider === "google"
      ? "https://oauth2.googleapis.com/token"
      : "https://login.microsoftonline.com/common/oauth2/v2.0/token";
    const body = new URLSearchParams({
      client_id: clientId,
      refresh_token: credential.refreshToken,
      grant_type: "refresh_token"
    });
    if (credential.provider === "microsoft") body.set("scope", this.scopes(credential.provider).join(" "));
    const token = await this.tokenRequest(tokenEndpoint, body);
    if (!token.access_token) throw oauthFailure("OAuth access token 刷新失败，请重新授权邮箱");
    return {
      ...credential,
      accessToken: token.access_token,
      refreshToken: token.refresh_token || credential.refreshToken,
      expiresAt: Date.now() + Math.max(60, Number(token.expires_in ?? 3600)) * 1000,
      scope: token.scope || credential.scope
    };
  }

  private async tokenRequest(endpoint: string, body: URLSearchParams): Promise<TokenResponse> {
    let response: Response;
    try {
      response = await fetch(endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body,
        signal: AbortSignal.timeout(15_000)
      });
    } catch (error) {
      throw oauthFailure(`OAuth 服务连接失败：${error instanceof Error ? error.message : String(error)}`, 503);
    }
    const token = await response.json().catch(() => ({})) as TokenResponse;
    if (!response.ok || token.error) {
      const message = token.error_description || token.error || `HTTP ${response.status}`;
      const status = token.error === "invalid_grant" ? 401 : response.status >= 500 ? 503 : 400;
      throw oauthFailure(`OAuth token 请求失败：${message}`, status);
    }
    return token;
  }

  private cleanupFlows(): void {
    const cutoff = Date.now() - flowLifetimeMs;
    for (const [id, flow] of this.flows) {
      if (flow.createdAt < cutoff) this.flows.delete(id);
    }
  }
}

export type { OAuthCredential };

File: src-tauri/src/oauth.rs
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use rand::{rngs::OsRng, RngCore};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::io::{ErrorKind, Read, Write};
use std::net::{TcpListener, TcpStream};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
use url::Url;

const OAUTH_TIMEOUT: Duration = Duration::from_secs(10 * 60);

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopOAuthCredential {
    version: u8,
    provider: String,
    email: String,
    display_name: String,
    client_id: String,
    access_token: String,
    refresh_token: String,
    expires_at: u64,
    scope: String,
}

#[derive(Debug, Deserialize)]
struct TokenResponse {
    access_token: Option<String>,
    refresh_token: Option<String>,
    expires_in: Option<u64>,
    scope: Option<String>,
    id_token: Option<String>,
    error: Option<String>,
    error_description: Option<String>,
}

struct OAuthProviderConfig {
    authorization_endpoint: &'static str,
    token_endpoint: &'static str,
    scopes: &'static [&'static str],
    redirect_host: &'static str,
}

fn provider_config(provider: &str) -> Result<OAuthProviderConfig, String> {
    match provider {
        "google" => Ok(OAuthProviderConfig {
            authorization_endpoint: "https://accounts.google.com/o/oauth2/v2/auth",
            token_endpoint: "https://oauth2.googleapis.com/token",
            scopes: &["openid", "email", "profile", "https://mail.google.com/"],
            redirect_host: "127.0.0.1",
        }),
        "microsoft" => Ok(OAuthProviderConfig {
            authorization_endpoint:
                "https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
            token_endpoint: "https://login.microsoftonline.com/common/oauth2/v2.0/token",
            scopes: &[
                "openid",
                "profile",
                "email",
                "offline_access",
                "https://outlook.office.com/IMAP.AccessAsUser.All",
                "https://outlook.office.com/SMTP.Send",
            ],
            redirect_host: "localhost",
        }),
        _ => Err("不支持的 OAuth 邮箱服务商".to_string()),
    }
}

fn random_base64_url(bytes: usize) -> String {
    let mut value = vec![0_u8; bytes];
    OsRng.fill_bytes(&mut value);
    URL_SAFE_NO_PAD.encode(value)
}

fn validate_client_id(value: String) -> Result<String, String> {
    let client_id = value.trim().to_string();
    if client_id.len() < 3 || client_id.len() > 512 || client_id.chars().any(char::is_whitespace) {
        return Err("OAuth Client ID 格式不正确".to_string());
    }
    Ok(client_id)
}

fn callback_response(stream: &mut TcpStream, success: bool) {
    let (title, message, color) = if success {
        (
            "授权已完成",
            "可以关闭此页面并返回 Mail Collector。",
            "#176b57",
        )
    } else {
        (
            "授权未完成",
            "请关闭此页面并返回 Mail Collector 重试。",
            "#b42318",
        )
    };
    let body = format!(
        "<!doctype html><meta charset=\"utf-8\"><title>Mail Collector</title><style>body{{font-family:system-ui;margin:48px;line-height:1.6;color:#1f2937}}main{{max-width:560px;margin:auto}}h1{{font-size:24px;color:{color}}}</style><main><h1>{title}</h1><p>{message}</p></main>"
    );
    let response = format!(
        "HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\nContent-Length: {}\r\nConnection: close\r\nCache-Control: no-store\r\n\r\n{}",
        body.len(),
        body
    );
    let _ = stream.write_all(response.as_bytes());
    let _ = stream.flush();
}

fn read_callback_request(stream: &mut TcpStream) -> Result<String, String> {
    stream
        .set_read_timeout(Some(Duration::from_secs(5)))
        .map_err(|error| error.to_string())?;
    let mut request = Vec::with_capacity(2048);
    let mut buffer = [0_u8; 2048];
    loop {
        let count = stream
            .read(&mut buffer)
            .map_err(|error| error.to_string())?;
        if count == 0 {
            break;
        }
        request.extend_from_slice(&buffer[..count]);
        if request.windows(4).any(|window| window == b"\r\n\r\n") {
            break;
        }
        if request.len() > 16 * 1024 {
            return Err("OAuth 回调请求过大".to_string());
        }
    }
    let request = String::from_utf8(request).map_err(|_| "OAuth 回调格式不正确".to_string())?;
    let target = request
        .lines()
        .next()
        .and_then(|line| line.split_whitespace().nth(1))
        .ok_or_else(|| "OAuth 回调请求不完整".to_string())?;
    Ok(target.to_string())
}

fn wait_for_callback(listener: TcpListener, expected_state: String) -> Result<String, String> {
    listener
        .set_nonblocking(true)
        .map_err(|error| error.to_string())?;
    let deadline = Instant::now() + OAUTH_TIMEOUT;
    loop {
        match listener.accept() {
            Ok((mut stream, _)) => {
                let target = read_callback_request(&mut stream)?;
                let callback = Url::parse(&format!("http://localhost{target}"))
                    .map_err(|_| "OAuth 回调地址格式不正确".to_string())?;
                let state = callback
                    .query_pairs()
                    .find(|(name, _)| name == "state")
                    .map(|(_, value)| value.into_owned())
                    .unwrap_or_default();
                if state != expected_state {
                    callback_response(&mut stream, false);
                    return Err("OAuth state 校验失败，请重新授权".to_string());
                }
                if let Some(error) = callback
                    .query_pairs()
                    .find(|(name, _)| name == "error_description" || name == "error")
                    .map(|(_, value)| value.into_owned())
                {
                    callback_response(&mut stream, false);
                    return Err(format!("授权未完成：{error}"));
                }
                let code = callback
                    .query_pairs()
                    .find(|(name, _)| name == "code")
                    .map(|(_, value)| value.into_owned())
                    .filter(|value| !value.is_empty())
                    .ok_or_else(|| "授权服务器没有返回授权码".to_string())?;
                callback_response(&mut stream, true);
                return Ok(code);
            }
            Err(error) if error.kind() == ErrorKind::WouldBlock => {
                if Instant::now() >= deadline {
                    return Err("OAuth 授权等待超时，请重新尝试".to_string());
                }
                std::thread::sleep(Duration::from_millis(50));
            }
            Err(error) => return Err(format!("OAuth 本地回调监听失败：{error}")),
        }
    }
}

fn jwt_payload(token: &str) -> Result<Value, String> {
    let payload = token
        .split('.')
        .nth(1)
        .ok_or_else(|| "OAuth 身份令牌格式不正确".to_string())?;
    let decoded = URL_SAFE_NO_PAD
        .decode(payload)
        .map_err(|_| "OAuth 身份令牌无法解析".to_string())?;
    serde_json::from_slice(&decoded).map_err(|_| "OAuth 身份令牌无法解析".to_string())
}

fn claim<'a>(payload: &'a Value, names: &[&str]) -> &'a str {
    for name in names {
        if let Some(value) = payload.get(name).and_then(Value::as_str) {
            if !value.trim().is_empty() {
                return value.trim();
            }
        }
    }
    ""
}

fn audience_matches(payload: &Value, client_id: &str) -> bool {
    match payload.get("aud") {
        Some(Value::String(value)) => value == client_id,
        Some(Value::Array(values)) => values.iter().any(|value| value.as_str() == Some(client_id)),
        _ => false,
    }
}

async fn exchange_code(
    provider: &str,
    config: &OAuthProviderConfig,
    client_id: &str,
    redirect_uri: &str,
    verifier: &str,
    nonce: &str,
    code: &str,
) -> Result<DesktopOAuthCredential, String> {
    let scope = config.scopes.join(" ");
    let mut form = vec![
        ("client_id", client_id.to_string()),
        ("code", code.to_string()),
        ("code_verifier", verifier.to_string()),
        ("redirect_uri", redirect_uri.to_string()),
        ("grant_type", "authorization_code".to_string()),
    ];
    if provider == "microsoft" {
        form.push(("scope", scope.clone()));
    }
    let response = Client::new()
        .post(config.token_endpoint)
        .timeout(Duration::from_secs(20))
        .form(&form)
        .send()
        .await
        .map_err(|error| format!("OAuth 服务连接失败：{error}"))?;
    let status = response.status();
    let token: TokenResponse = response
        .json()
        .await
        .map_err(|_| "OAuth token 响应格式不正确".to_string())?;
    if !status.is_success() || token.error.is_some() {
        let message = token
            .error_description
            .or(token.error)
            .unwrap_or_else(|| format!("HTTP {status}"));
        return Err(format!("OAuth token 请求失败：{message}"));
    }
    let access_token = token
        .access_token
        .filter(|value| !value.is_empty())
        .ok_or_else(|| "授权服务器没有返回 access token".to_string())?;
    let refresh_token = token
        .refresh_token
        .filter(|value| !value.is_empty())
        .ok_or_else(|| "授权服务器没有返回 refresh token，请撤销旧授权后重试".to_string())?;
    let id_token = token
        .id_token
        .filter(|value| !value.is_empty())
        .ok_or_else(|| "授权服务器没有返回身份令牌".to_string())?;
    let payload = jwt_payload(&id_token)?;
    if !audience_matches(&payload, client_id) {
        return Err("OAuth 身份令牌的客户端不匹配".to_string());
    }
    if claim(&payload, &["nonce"]) != nonce {
        return Err("OAuth 身份令牌 nonce 校验失败".to_string());
    }
    let issuer = claim(&payload, &["iss"]);
    if provider == "google"
        && issuer != "https://accounts.google.com"
        && issuer != "accounts.google.com"
    {
        return Err("Google OAuth 身份令牌签发方不正确".to_string());
    }
    if provider == "microsoft" && !issuer.starts_with("https://login.microsoftonline.com/") {
        return Err("Microsoft OAuth 身份令牌签发方不正确".to_string());
    }
    let email = claim(&payload, &["email", "preferred_username", "upn"]).to_string();
    if !email.contains('@') {
        return Err("无法从 OAuth 授权结果识别邮箱地址".to_string());
    }
    if provider == "google" && payload.get("email_verified") == Some(&Value::Bool(false)) {
        return Err("Google 账户邮箱地址尚未验证".to_string());
    }
    let display_name = {
        let name = claim(&payload, &["name"]);
        if name.is_empty() {
            email.split('@').next().unwrap_or(&email).to_string()
        } else {
            name.to_string()
        }
    };
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|_| "系统时间不正确".to_string())?
        .as_millis() as u64;
    Ok(DesktopOAuthCredential {
        version: 1,
        provider: provider.to_string(),
        email,
        display_name,
        client_id: client_id.to_string(),
        access_token,
        refresh_token,
        expires_at: now + token.expires_in.unwrap_or(3600).max(60) * 1000,
        scope: token.scope.unwrap_or(scope),
    })
}

#[tauri::command]
pub async fn authorize_mail_provider(
    provider: String,
    client_id: String,
) -> Result<DesktopOAuthCredential, String> {
    let config = provider_config(&provider)?;
    let client_id = validate_client_id(client_id)?;
    let listener = TcpListener::bind("127.0.0.1:0")
        .map_err(|error| format!("无法启动 OAuth 本地回调：{error}"))?;
    let port = listener
        .local_addr()
        .map_err(|error| error.to_string())?
        .port();
    let redirect_uri = format!("http://{}:{port}", config.redirect_host);
    let state = random_base64_url(32);
    let nonce = random_base64_url(24);
    let verifier = random_base64_url(48);
    let challenge = URL_SAFE_NO_PAD.encode(Sha256::digest(verifier.as_bytes()));
    let mut authorization_url =
        Url::parse(config.authorization_endpoint).map_err(|error| error.to_string())?;
    {
        let mut query = authorization_url.query_pairs_mut();
        query
            .append_pair("client_id", &client_id)
            .append_pair("response_type", "code")
            .append_pair("redirect_uri", &redirect_uri)
            .append_pair("response_mode", "query")
            .append_pair("scope", &config.scopes.join(" "))
            .append_pair("state", &state)
            .append_pair("nonce", &nonce)
            .append_pair("code_challenge", &challenge)
            .append_pair("code_challenge_method", "S256")
            .append_pair(
                "prompt",
                if provider == "google" {
                    "consent select_account"
                } else {
                    "select_account"
                },
            );
        if provider == "google" {
            query.append_pair("access_type", "offline");
        }
    }
    crate::open_external_url(authorization_url.to_string())?;
    let callback_state = state.clone();
    let code =
        tauri::async_runtime::spawn_blocking(move || wait_for_callback(listener, callback_state))
            .await
            .map_err(|error| format!("OAuth 本地回调任务失败：{error}"))??;
    exchange_code(
        &provider,
        &config,
        &client_id,
        &redirect_uri,
        &verifier,
        &nonce,
        &code,
    )
    .await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn provider_configuration_keeps_native_loopback_redirects() {
        assert_eq!(
            provider_config("google").unwrap().redirect_host,
            "127.0.0.1"
        );
        assert_eq!(
            provider_config("microsoft").unwrap().redirect_host,
            "localhost"
        );
        assert!(provider_config("other").is_err());
    }

    #[test]
    fn client_id_rejects_whitespace_and_empty_values() {
        assert_eq!(
            validate_client_id(" valid-client ".to_string()).unwrap(),
            "valid-client"
        );
        assert!(validate_client_id("".to_string()).is_err());
        assert!(validate_client_id("invalid client".to_string()).is_err());
    }
}
